File: src/lib/utils/retry.ts
import { logger } from './logger';

/**
 * Retry a function with exponential backoff
 */
export async function withRetry<T>(
  fn: () => Promise<T>,
  maxRetries = 3,
  baseDelay = 1000
): Promise<T> {
  let lastError: Error | undefined;

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error instanceof Error ? error : new Error(String(error));

      if (attempt === maxRetries) break;

      // Exponential backoff with jitter
      const delay = baseDelay * Math.pow(2, attempt) + Math.random() * 500;
      logger.warn('Retrying after failure', { attempt: attempt + 1, maxRetries, delayMs: Math.round(delay), error: lastError.message });
      await new Promise(resolve => setTimeout(resolve, delay));
    }
  }

  throw lastError;
}

/**
 * Retry a function that returns null on failure (treating null as a retriable failure)
 */
export async function withRetryNullable<T>(
  fn: () => Promise<T | null>,
  maxRetries = 2,
  baseDelay = 1000
): Promise<T | null> {
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    const result = await fn();
    if (result !== null) return result;

    if (attempt === maxRetries) return null;

    const delay = baseDelay * Math.pow(2, attempt) + Math.random() * 500;
    logger.warn('Retrying after null result', { attempt: attempt + 1, maxRetries, delayMs: Math.round(delay) });
    await new Promise(resolve => setTimeout(resolve, delay));
  }

  return null;
}

File: internal/util/retry.go
package util

import (
	"context"
	"math/rand"
	"time"
)

// WithRetry executes fn repeatedly with exponential backoff and jitter.
// maxRetries is the number of retries after the initial attempt (total attempts = maxRetries+1).
func WithRetry[T any](ctx context.Context, fn func(context.Context) (T, error), maxRetries int, baseDelay time.Duration) (T, error) {
	var zero T
	for i := 0; ; i++ {
		result, err := fn(ctx)
		if err == nil {
			return result, nil
		}
		if i >= maxRetries {
			return zero, err
		}
		delay := baseDelay * (1 << i) // exponential: baseDelay * 2^i
		delay += time.Duration(rand.Float64() * float64(500*time.Millisecond))
		Warn("Retry attempt", "attempt", i+1, "retry_delay", delay, "error", err)
		select {
		case <-ctx.Done():
			return zero, ctx.Err()
		case <-time.After(delay):
		}
	}
}

// WithRetryPointer is for functions that signal retryable failure by returning (nil, nil).
// Non-nil results are returned immediately; non-nil errors are bubbled immediately.
func WithRetryPointer[T any](ctx context.Context, fn func(context.Context) (*T, error), maxRetries int, baseDelay time.Duration) (*T, error) {
	for i := 0; ; i++ {
		res, err := fn(ctx)
		if err != nil {
			return nil, err
		}
		if res != nil {
			return res, nil
		}
		if i >= maxRetries {
			return nil, nil
		}
		delay := baseDelay * (1 << i)
		delay += time.Duration(rand.Float64() * float64(500*time.Millisecond))
		Warn("Retry attempt (pointer)", "attempt", i+1, "retry_delay", delay)
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(delay):
		}
	}
}
